test/openssl/test_pkey.rb

```diff
@@ -244,4 +244,210 @@ def test_to_text
+                                                      priv: source.private_key.to_i)
+        assert_instance_of OpenSSL::PKey::EC, new_key
+        assert_equal source.group.curve_name, new_key.group.curve_name
+        assert_equal source.private_key, new_key.private_key
+        assert_equal source.public_key, new_key.public_key
+      end
+    end
+
+    def test_s_from_parameters_ec_pub_given_as_integer
+      e = assert_raise(TypeError) { OpenSSL::PKey.from_parameters("EC", { group: "prime256v1", pub: 12345 }) }
+      assert_equal "no implicit conversion of Integer into String", e.message
+    end
+
+    def test_s_from_parameters_ec_with_invalid_parameter
+      e = assert_raise(OpenSSL::PKey::PKeyError) { OpenSSL::PKey.from_parameters("EC", invalid: 1234) }
+      assert_match(/Invalid parameter "invalid"/, e.message)
+    end
+
+    def test_s_from_parameters_dsa_with_all_supported_parameters
+      source  = OpenSSL::PKey::DSA.generate(2048)
+
+      new_key =  OpenSSL::PKey.from_parameters("DSA", pub: source.params["pub_key"],
+                                                      priv: source.params["priv_key"],
+                                                      p: source.params["p"],
+                                                      q: source.params["q"],
+                                                      g: source.params["g"])
+
+      assert_instance_of OpenSSL::PKey::DSA, new_key
+      assert_equal source.params, new_key.params
+    end
+
+    def test_s_from_parameters_dsa_with_gem_specific_keys
+      source  = OpenSSL::PKey::DSA.generate(2048)
+
+      new_key =  OpenSSL::PKey.from_parameters("DSA", source.params)
+
+      assert_equal source.params, new_key.params
+    end
+
+    def test_s_from_parameters_dsa_with_invalid_parameter
+      e = assert_raise(OpenSSL::PKey::PKeyError) { OpenSSL::PKey.from_parameters("DSA", invalid: 1234) }
+      assert_match(/Invalid parameter "invalid"/, e.message)
+    end
+
+    def test_s_from_parameters_dh_with_all_supported_parameters
+      source  = OpenSSL::PKey::DH.generate(2048)
+
+      new_key =  OpenSSL::PKey.from_parameters("DH", source.params)
+
+      assert_instance_of OpenSSL::PKey::DH, new_key
+      assert_equal source.params, new_key.params
+    end
+
+    def test_s_from_parameters_dh_with_invalid_parameter
+      e = assert_raise(OpenSSL::PKey::PKeyError) { OpenSSL::PKey.from_parameters("DH", invalid: 1234) }
+      assert_match(/Invalid parameter "invalid"/, e.message)
+    end
+
+    def test_s_from_parameters_ed25519
+      # Ed25519 is not FIPS-approved.
+      omit_on_fips
+
+      key = OpenSSL::PKey.from_parameters("ED25519", pub: "\xD0\x8E\xA8\x96\xB6Fbi{$k\xAC\xB8\xA2V\xF4n\xC3\xD06}R\x8A\xE6I\xA7r\xF6D{W\x84")
+      assert_instance_of OpenSSL::PKey::PKey, key
+      assert_equal "-----BEGIN PUBLIC KEY-----\nMCowBQYDK2VwAyEA0I6olrZGYml7JGusuKJW9G7D0DZ9UormSady9kR7V4Q=\n-----END PUBLIC KEY-----\n", key.public_to_pem
```

Comment:
I think you can use Ruby's heredoc syntax for a better readability. You can refer to other parts using the heredoc in the `test/openssl/test_pkey.rb`.

Reply:
Adjusted to have the expected public presented in heredoc. Looks way prettier